T/48

f1ames opened the pull request:
Feature: Introduced `InputCommand` used for typing handling. Closes #48.

Reinmar commented on src/inputcommand.js:
```diff
@@ -0,0 +1,100 @@
+		 * @protected
+		 * @member {typing.ChangeBuffer} #_buffer
+		 */
+		this._buffer = new ChangeBuffer( editor.document, editor.config.get( 'typing.undoStep' ) || 20 );
```
This should be given to the command as an option, not taken by it from the editor. A class like a command should depend on the world in as little details as possible.

Reinmar commented on src/input.js:
```diff
@@ -72,29 +55,28 @@ export default class Input extends Plugin {
 	 *
-	 * @param {Array.<engine.view.Document~MutatatedText|engine.view.Document~MutatatedChildren>} mutations
+	 * @param {Array.<module:engine/view/document~MutatatedText|module:engine/view/document~MutatatedChildren>} mutations
+	 * @param {module:engine/controller/editingcontroller~EditingController} viewSelection
```
Wrong type.

Reinmar commented on src/inputcommand.js:
```diff
@@ -0,0 +1,100 @@
+	 * @param {String} [options.text=''] Text to be inserted.
+	 * @param {module:engine/model/range~Range} [options.range=null] Range in which the text is inserted. Defaults
+	 * to first range in the current selection.
+	 * @param {module:engine/model/position~Position} [options.selectionAnchor] Selection anchor which will be used
```
This command should not need to accept the selection anchor (I don't even know how it's used). It should calculate the result selection position based on the given target range to remove and text which it's inserting. At least... I hope that it can be done this way ;>

f1ames replied:
For simple case it works as you described. But there are cases like e.g. spell checking and probably it was the reason that [modelSelectionPosition](https://github.com/ckeditor/ckeditor5-typing/blob/master/src/input.js#L207) was introduced at the first place.
If there is text like `athat` and spell checking is used to replaced it with `that`, based on mutation there is range like `[a]that` and inserted text is empty. I am not sure if it is possible to calculate proper selection position based on that (I assume after replacing the position should be after replaced/corrected word)?
I may propose changing the option name to e.g. `targetPosition` which may be a little more meaningful, unless ofc  there is a way to correctly calculate selection in those cases?

Reinmar replied:
If you replace `athat` with `that` and according to the engine, the selection should be after "that", then the passes `options.range` should be `[athat]` and `options.text = 'that'`. The input command will remove the whole "athat", insert "that" and place the selection after "that". You don't need `options.selectionAnchor` for that.

I understand that this may not be fully intuitive at first, but think about this from OT perspective. If you're fixing the word's "athat" spelling you're replacing it with "that". The whole word, not just a single letter. This may be crucial when resolving some conflicting changes and, in general, is more semantical.

Hence, try to get rid of `selectionAnchor`. I wrote before that I'm not sure if it's doable, though, because I haven't checked input feature's implementation. Why did you implement it in the first place? Was it already strictly required by what the mutation observer produces?

f1ames replied:
Ok, I was relying too much on what information we already have calculated based on mutations.

For case described above the mutation provided by mutation observer is:

```
newText: "foo that bar"
node: Text
oldText: "foo athat bar"
type: "text"
```

I assume based only on this information it is not possible to reliable calculate the range and text which will be passed to `InputCommand`, unless we may assume that the range (and text) will consist of all words which were modified by mutation. If not I think the mutation observer should be modified to provide some additional information in mutations. 

The other spellchecking cases to consider are:
`Foo hous a` -> `Foo house a`
`Foo hous e` -> `Foo house`
`Foo athat`   -> `Food that`
`Foo xhat` -> `Foo chat`

f1ames replied:
Hmmm, forgot about `viewSelection` which is passed along with mutations, will see if it will help solve this one :)

Reinmar commented on src/input.js:
```diff
@@ -211,20 +184,11 @@ class MutationHandler {
-		}
+		this.editor.execute( 'input', {
+			text: mutation.newText.substr( firstChangeAt, insertions ),
+			range: deletions > 0 ? ModelRange.createFromPositionAndShift( modelPos, deletions ) : null,
```
Please move the consts out of this object literal. Those variables can make the code more readable (and easier to debug) thanks to their names.

Reinmar commented on src/inputcommand.js:
```diff
@@ -0,0 +1,100 @@
+/**
+ * The input command. Used by the {@link module:typing/input~Input input feature} to handle typing.
+ *
+ * @extends core.command.Command
```
Old notation. Now should be `@extends {module:blebleble~Command}`.

Reinmar commented on src/inputcommand.js:
```diff
@@ -0,0 +1,100 @@
+		 * Typing's change buffer used to group subsequent changes into batches.
+		 *
+		 * @protected
+		 * @member {typing.ChangeBuffer} #_buffer
```
Old type notation as well.

Reinmar commented on src/inputcommand.js:
```diff
@@ -0,0 +1,100 @@
+	 *
+	 * @param {Object} [options] The command options.
+	 * @param {String} [options.text=''] Text to be inserted.
+	 * @param {module:engine/model/range~Range} [options.range=null] Range in which the text is inserted. Defaults
```
I think that the range could default to first selection's range if not passed. Will be pretty useful.

f1ames replied:
It does, it is stated in the comment: _Range in which the text is inserted. Defaults to first range in the current selection._ However, it is no stated in the param specs, should it be like `[options.range] plus description` or something like `[options.range=document.selection.getFirstRange()]`?

Reinmar replied:
Can be `[options.range] plus description`. Can also be the latter. But definitely not `=null` :D

f1ames replied:
Makes perfect sense, I'll go with first one.

f1ames commented:
Pushed changes suggested by @Reinmar. There are two things which may require further discussion or some changes, see: https://github.com/ckeditor/ckeditor5-typing/pull/71#discussion_r101022539 and https://github.com/ckeditor/ckeditor5-typing/pull/71#discussion_r101020183.

Reinmar commented on src/input.js:
```diff
@@ -107,45 +108,45 @@ class MutationHandler {
+		 * The change buffer;
 		 *
-		 * @member {Number} #insertedCharacterCount
+		 * @type {module:typing/changebuffer~ChangeBuffer} #buffer
```
Should be `@member`.

f1ames commented:
Corrected the docs.

Reinmar commented:
I commented on the two aspects you mentioned. The most important bit is to try to avoid the 3rd option.

Also, you'll need to resolve conflicts now.

PS. You may not know yet  – you shouldn't rebase a branch once it's after any review because this will break comments on GH. So merge master to it.

f1ames commented:
I moved `buffer` back to `InputCommand` as I earlier misunderstood this comment https://github.com/ckeditor/ckeditor5-typing/pull/71#pullrequestreview-21564067. Updated unit tests accordingly.
We decide that `InputCommand` would have optional parameter/option `resultPosition`, if given, after text processing the selection is set using `resultPosition`.

Reinmar commented:
I fixed the API docs a bit.

I didn't like one thing – the fact that `options.text` is optional. Executing the command without the text doesn't make sense because it doesn't do anything.

Another thing is that the range doesn't have to be checked too. I don't see a case today in which selection could not have any range in it. The document selection will return so called "default range" and if you provide any custom selection, with no range inside, let the code throw. At least we'll know the situation occurs so we'll be able to fix it. Now, it's a silent error.

Reinmar commented:
Of course, let's do that unless we already see that there may be no range or no text :D. Let's then think how to handle this situation.

f1ames commented:
> Executing the command without the text doesn't make sense because it doesn't do anything.

Well, it does not look like making much sense indeed. But without checking if text was passed and passing non-collapsed range at the same time to the command will result in removing the content within the range without inserting anything. So we should decide if it is an acceptable behavior or some error should be thrown if text was not passed.
It is the same behavior as passing `text = ''`, I am not sure if we should differentiate this two cases or assume the it should behaves the same.

Reinmar commented on src/inputcommand.js:
```diff
@@ -0,0 +1,101 @@
+		const range = options.range || doc.selection.getFirstRange();
+		const resultPosition = options.resultPosition;
+
+		let textInsertions = 0;
```
Is it needed?

f1ames commented:
Made `options.text` optional and defaults to `''` as we discussed earlier.
